// unplugin-vue-router/dist/chunk-Y3YJJQFJ.mjs
import {
  getFileBasedRouteName,
  isArray
} from "./chunk-L2OPP4OY.mjs";

// src/options.ts
import { isPackageExists } from "local-pkg";
import { resolve } from "pathe";
var DEFAULT_OPTIONS = {
  extensions: [".vue"],
  exclude: [],
  routesFolder: [{ src: "src/pages" }],
  routeBlockLang: "json5",
  getRouteName: getFileBasedRouteName,
  dataFetching: false,
  importMode: "async",
  root: process.cwd(),
  dts: isPackageExists("typescript"),
  logs: false,
  _inspect: false
};
function normalizeRoutesFolderOption(routesFolder) {
  return (isArray(routesFolder) ? routesFolder : [routesFolder]).map(
    (routeOption) => typeof routeOption === "string" ? { src: routeOption } : routeOption
  );
}
function resolveOptions(options) {
  const root = options.root || DEFAULT_OPTIONS.root;
  const routesFolder = normalizeRoutesFolderOption(
    options.routesFolder || DEFAULT_OPTIONS.routesFolder
  ).map((routeOption) => ({
    ...routeOption,
    src: resolve(root, routeOption.src)
  }));
  return {
    ...DEFAULT_OPTIONS,
    ...options,
    routesFolder
  };
}

export {
  DEFAULT_OPTIONS,
  resolveOptions
};
